libmscore/connector.cpp

```diff
@@ -0,0 +1,447 @@
+//   ConnectorInfo
+//---------------------------------------------------------
+
+ConnectorInfo::ConnectorInfo(const Element* current, int track, Fraction frac)
+   : _current(current), _currentInfo(PointInfo::absolute())
+      {
+      if (!current)
+            qFatal("ConnectorInfo::ConnectorInfo(): invalid argument: %p", current);
+      // It is not always possible to determine the track number correctly from
+      // the current element (for example, in case of a Segment).
+      // If the caller does not know the track number and passes -1
+      // it may be corrected later.
+      if (track >= 0)
+            _currentInfo.setTrack(track);
+      if (frac >= 0)
+            _currentInfo.setFrac(frac);
+      }
+
+//---------------------------------------------------------
+//   ConnectorInfo
+//---------------------------------------------------------
+
+ConnectorInfo::ConnectorInfo(const PointInfo& currentInfo)
+   : _currentInfo(currentInfo)
+      {}
+
+//---------------------------------------------------------
+//   ConnectorInfo::updatePointInfo
+//---------------------------------------------------------
+
+void ConnectorInfo::updatePointInfo(const Element* e, PointInfo& i, bool clipboardmode)
+      {
+      i.fillForElement(e, clipboardmode);
+      }
+
+//---------------------------------------------------------
+//   ConnectorInfo::updateCurrentInfo
+//---------------------------------------------------------
+
+void ConnectorInfo::updateCurrentInfo(bool clipboardmode) {
+      if (!currentUpdated() && _current)
+            updatePointInfo(_current, _currentInfo, clipboardmode);
+      setCurrentUpdated(true);
+}
+
+//---------------------------------------------------------
+//   ConnectorInfo::connect
+//---------------------------------------------------------
+
+bool ConnectorInfo::connect(ConnectorInfo* other)
+      {
+      if (!other || (this == other))
+            return false;
+      if (_type != other->_type)
+            return false;
+      if (!_prev && hasPrevious() && !other->_next && other->hasNext()) {
+            if ((_prevInfo == other->_currentInfo)
+               && (_currentInfo == other->_nextInfo)
+               ) {
+                  _prev = other;
+                  other->_next = this;
+                  return true;
+                  }
+            }
+      if (!_next && hasNext() && !other->_prev && other->hasPrevious()) {
+            if ((_nextInfo == other->_currentInfo)
+               && (_currentInfo == other->_prevInfo)
+               ) {
+                  _next = other;
+                  other->_prev = this;
+                  return true;
+                  }
+            }
+      return false;
+      }
+
+//---------------------------------------------------------
+//   ConnectorInfo::forceConnect
+//---------------------------------------------------------
+
+void ConnectorInfo::forceConnect(ConnectorInfo* other)
+      {
+      if (!other || (this == other))
```

Comment:
But those `Q_ASSERT`s are stripped from release builds, I guess?
That says also the [developers handbook](https://musescore.org/en/handbook/developers-handbook/finding-your-way-around/musescore-coding-rules#Qt-Debugging-constructs):
`Q_ASSERT(bool test) should be used for errors for which musescore can still continue on. Q_ASSERT does nothing when built in release, but will halt execution when built in debug.`

This check covers cases which are hardly acceptable for MuseScore runtime: a null pointer would lead to a crash and `this == other` would cause an eternal loop when something in the code would try to find an endpoint of such a connector. Perhaps there could be a warning added but I would not replace this check with `Q_ASSERT`.

Reply:
Ok cool. I just wondered whether you might be able to catch this before the function is called, but if you can't be sure of doing that then this is fine.